ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java

```diff
@@ -294,8 +293,9 @@ static long writeTo(ByteBuf buf, WriteOption[] options, DataStream stream) {
     final DataChannel channel = stream.getDataChannel();
     long byteWritten = 0;
     for (ByteBuffer buffer : buf.nioBuffers()) {
+      final ReferenceCountedObject<ByteBuffer> wrapped = ReferenceCountedObject.wrap(buffer, buf::retain, buf::release);
```

Comment:
Is it possible that for a buffer, buf retain and release are both called while no other retain invocation is done? In such a case buf would get deallocated and can not be used by other buffers.

Reply:
@lokeshj1703 , thanks for reviewing this.

Ratis would have buf.retain() called at least once before passing it to StateMachine via write(wrapped).  After write(wrapped), Ratis will release it.  StateMachine implementation may choose to further retain it.  In such case, the buf will be released until both Ratis and the StateMachine have released it.  If StateMachine don't retain it, the buf will be released by Ratis as usual.

Reply:
I see. I was thinking of a case where there are more than one buffers. For first buffer, buf retain and release is called by Ratis and reference count for buf would reach zero. According to ByteBuf#release documentation, if ref count goes to zero buf will be released. Based on this after one buffer is processed, buf will be released and subsequent buffer processing would fail?

Reply:
When there are more than one ByteBuffer objects backed by a ByteBuf, Ratis retains the ByteBuf first and then calls write(ByteBuffer) for each ByteBuffer objects.  Once all the write calls have completed, it releases the ByteBuf.

In other words, whenever Ratis calling write(ByteBuffer), it makes sure that the backed ByteBuf has reference count >= 1.  Logically, it is equivalent to do the following for each write(..) call.
1. Ratis retains.
2. Ratis calls write(..).
3. Ratis releases.

Reply:
Got it. Thanks for the explanation @szetszwo!